### DISCUSSION TITLE
Embedded emojis within text

### DISCUSSION THREAD

**Discussion by avisra** [SixLabors/Fonts#413] [Q&A] [answered]
Here's a fun one :)

The text is accept comes from end-users. Occasionally, they like to slap an emoji along with some text. The text itself might be Font A - but when SixLabors attempts to render it - the emoji is missing because its trying to render everything with Font A. Is there a way to update it so it can use two fonts within a single TextOptions? Or what is the best way to handle this?

**Comment by tocsoft** [comment]
you need to specify `textOptions.FallbackFontFamilies` to emojii font with ttf outlines.

We don't have a great sample of this but this line https://github.com/SixLabors/Fonts/blob/fec131d9b6b1461752422f3c757e1d59779a1106/samples/DrawWithImageSharp/Program.cs#L89 is doing just that (the helper `RenderText`) is just creating the text options and then rendering it out.

**Reply by tocsoft** [reply]
Fallback fonts can be used for any glyphs missing from the primary font (not just emojii), the missing glyph will be looked up from the primary font first, then it will be looked up in each of the fallback fonts in the order provided, stopping at the first hit.

Emoji are the most obvious visible use case for English speakers for fallback fonts but places like international fonts with Kanji or Farsi glyphs are also prime examples where its useful to have a fallback... basically anywhere you might be mixing different character sets etc.

**Reply by tocsoft** [reply]
in fact it looks like our docs have a rough example too https://docs.sixlabors.com/articles/fonts/gettingstarted.html#expanded-example

**Reply by avisra** [reply]
Thanks a bunch for sharing all of this! I have tried it, but it is still not rendering anything for me - which is strange. I am primarily using DrawText and GenerateGlyphs. Will keep digging a bit.

The font is this one: https://fonts.google.com/noto/specimen/Noto+Color+Emoji

**Reply by avisra** [reply]
I switched over to Twemoji.Mozilla and it started working immediately. There must be something with the Noto Color Emoji that causes things to fail...

However, my code is still failing under one scenario with Twemoji.Mozilla - 
```
                        PathBuilder pathBuilder = new PathBuilder();
                        pathBuilder.SetOrigin(new PointF(xPos, 0));

                        pathBuilder.AddCubicBezier(new PointF(0, squareSize * (float).10), new PointF((int)(squareSize * .25), 0), new PointF((int)(squareSize * .75), 0), new PointF(squareSize, squareSize * (float).10));

                        IPath path = pathBuilder.Build();
                        var textOptions = new TextOptions(playerNameFont)
                        {
                            WrappingLength = path.ComputeLength(),
                            VerticalAlignment = VerticalAlignment.Top,
                            HorizontalAlignment = HorizontalAlignment.Left,
                            TextAlignment = TextAlignment.Center,
                            KerningMode = KerningMode.Auto,
                            Origin = new Point(0, 0),
                            FeatureTags = new List<Tag>() { FeatureTags.Kerning },
                            FallbackFontFamilies = new[] { emojiFamily.Value }
                        };

                        var glyphs = TextBuilder.GenerateGlyphs(playerName.Value, path, textOptions);

                        image.Mutate(ctx => ctx.Fill(new DrawingOptions() { ShapeOptions = new ShapeOptions { IntersectionRule = IntersectionRule.NonZero } }, playerNameColor.Color, glyphs));
```

This causes the emojis to be colored in solid black like the rest of the text. How can I prevent the emojis from being painted black?

**Reply by tocsoft** [reply, marked as answer]
You will want to move over to our updated `RichTextOptions` type in place of the base `TextOptions` which adds support for a `Path` property to render along used while rendering the text along with providing per span overrides often brush/pen used for part of the text. (`RichTextRun`)

I've just opened a PR to update our [samples](https://github.com/SixLabors/samples) to cover your use case.  https://github.com/SixLabors/Samples/pull/37 which produces an output like this
![image](https://github.com/user-attachments/assets/6ba34a8c-e889-4ec7-bbc4-fd2dc4dc82c5)


A direct link to the raw code is https://github.com/SixLabors/Samples/blob/6168903c095f16afe5816c52c34b103f2fba6e77/ImageSharp/DrawingTextAlongAPath/Program.cs

**Reply by avisra** [reply]
This worked wonderfully. Thank you!!
